File: core/papers/flashcards.go
// Package papers — flashcards model, parser, renderer and scheduler.
//
// The study-paper skill writes a deck as `flashcards.md` (see
// skills/study-paper/assets/flashcards.md): a plain Markdown document with a
// card table (| ID | Front | Back | Anchor | Tag | Stage |) and an append-only
// review log (| ID | Date | Grade | Next due |). This file holds the pure
// logic over that text — parse, render, and the fixed interval scheduler — with
// no I/O. The writer layer (writer.go) owns the atomic, containment-checked
// persistence; the backend exposes a mutation RPC on top of it.
//
// Everything here mirrors the frontend's lib/flashcards.ts and
// lib/spacedRepetition.ts so the review a reader sees and the review the
// backend records agree.

package papers

import (
	"errors"
	"fmt"
	"strings"
	"time"
)

// FlashcardFileName is the deck artifact's file name (inside a paper's dir).
const FlashcardFileName = "flashcards.md"

// FlashcardStage is a card's lifecycle stage.
type FlashcardStage string

const (
	// StageNew: never reviewed.
	StageNew FlashcardStage = "new"
	// StageLearning: reviewed but not yet past the first interval.
	StageLearning FlashcardStage = "learning"
	// StageReview: promoted past the first interval.
	StageReview FlashcardStage = "review"
)

// Grade is a self-graded review outcome.
type Grade string

const (
	// GradeAgain: missed — reset to the first interval.
	GradeAgain Grade = "again"
	// GradeHard: repeat at the same interval.
	GradeHard Grade = "hard"
	// GradeGood: promote to the next interval.
	GradeGood Grade = "good"
	// GradeEasy: promote to the next interval.
	GradeEasy Grade = "easy"
)

// Card is one atomic question/answer card.
type Card struct {
	ID     string         `json:"id"`
	Front  string         `json:"front"`
	Back   string         `json:"back"`
	Anchor string         `json:"anchor"`
	Tag    string         `json:"tag"`
	Stage  FlashcardStage `json:"stage"`
}

// ReviewEntry is one row of the review log.
type ReviewEntry struct {
	ID      string `json:"id"`
	Date    string `json:"date"`
	Grade   Grade  `json:"grade,omitempty"`
	NextDue string `json:"next_due,omitempty"`
}

// Deck is a parsed deck: the ordered cards plus the append-only review log.
type Deck struct {
	Cards   []Card        `json:"cards"`
	Reviews []ReviewEntry `json:"reviews"`
}

// NormalizeStage folds a raw stage token to its canonical value; an unknown or
// blank token maps to StageNew.
func NormalizeStage(raw string) FlashcardStage {
	switch strings.ToLower(cleanLine(raw)) {
	case "learning", "learn":
		return StageLearning
	case "review", "reviewing", "graduated":
		return StageReview
	default:
		return StageNew
	}
}

// NormalizeGrade folds a raw grade token to its canonical value; an unknown or
// blank token maps to "" (an ungraded review row).
func NormalizeGrade(raw string) Grade {
	switch strings.ToLower(cleanLine(raw)) {
	case "again":
		return GradeAgain
	case "hard":
		return GradeHard
	case "good":
		return GradeGood
	case "easy":
		return GradeEasy
	default:
		return ""
	}
}

// ---------------------------------------------------------------------------
// Column detection (shared with the frontend parser)
// ---------------------------------------------------------------------------

func headerMatches(header []string, names ...string) bool {
	return columnIndex(header, names...) >= 0
}

// isCardsHeader reports whether a table header is the deck's card table.
func isCardsHeader(header []string) bool {
	return headerMatches(header, "front", "question", "prompt") &&
		headerMatches(header, "back", "answer")
}

// isReviewHeader reports whether a table header is the review log.
func isReviewHeader(header []string) bool {
	return headerMatches(header, "grade", "rating", "result") || headerMatches(header, "next", "due")
}

// ---------------------------------------------------------------------------
// Parsing
// ---------------------------------------------------------------------------

// ParseFlashcards parses a flashcards.md document into its deck. Best-effort
// and total: a document with no recognizable table yields an empty deck, and a
// row carrying neither a prompt nor an answer (the template's placeholder) is
// skipped.
func ParseFlashcards(content string) Deck {
	var deck Deck
	for _, t := range parseTables(content) {
		if len(deck.Cards) == 0 && isCardsHeader(t.Header) {
			deck.Cards = parseCards(t)
			continue
		}
		if len(deck.Reviews) == 0 && isReviewHeader(t.Header) {
			deck.Reviews = parseReviews(t)
		}
	}
	return deck
}

func parseCards(t mdTable) []Card {
	idIdx := colOr(t.Header, 0, "id")
	frontIdx := colOr(t.Header, 0, "front", "question", "prompt")
	backIdx := colOr(t.Header, 1, "back", "answer")
	anchorIdx := columnIndexExcept(t.Header, []int{idIdx, frontIdx, backIdx}, "anchor", "source", "location", "ref", "where", "page")
	// Resolve Stage before Tag: "stage" contains the substring "tag", so a
	// tag-first pick would misclaim a Tag-less deck's Stage column.
	stageIdx := columnIndexExcept(t.Header, []int{idIdx, frontIdx, backIdx, anchorIdx}, "stage", "state", "status")
	tagIdx := columnIndexExcept(t.Header, []int{idIdx, frontIdx, backIdx, anchorIdx, stageIdx}, "tag", "topic", "concept", "label")
	var out []Card
	for _, row := range t.Rows {
		front := cellAt(row, frontIdx)
		back := cellAt(row, backIdx)
		if front == "" && back == "" {
			continue
		}
		out = append(out, Card{
			ID:     cellAt(row, idIdx),
			Front:  front,
			Back:   back,
			Anchor: cellAt(row, anchorIdx),
			Tag:    cellAt(row, tagIdx),
			Stage:  NormalizeStage(cellAt(row, stageIdx)),
		})
	}
	return out
}

func parseReviews(t mdTable) []ReviewEntry {
	idIdx := colOr(t.Header, 0, "id")
	dateIdx := columnIndexExcept(t.Header, []int{idIdx}, "date", "when")
	gradeIdx := columnIndexExcept(t.Header, []int{idIdx, dateIdx}, "grade", "rating", "result")
	dueIdx := columnIndexExcept(t.Header, []int{idIdx, dateIdx, gradeIdx}, "next", "due")
	var out []ReviewEntry
	for _, row := range t.Rows {
		id := cellAt(row, idIdx)
		date := cellAt(row, dateIdx)
		grade := cellAt(row, gradeIdx)
		due := cellAt(row, dueIdx)
		if id == "" && date == "" && grade == "" && due == "" {
			continue
		}
		out = append(out, ReviewEntry{ID: id, Date: date, Grade: NormalizeGrade(grade), NextDue: due})
	}
	return out
}

// columnIndexExcept is columnIndex over the header cells whose index is not in
// exclude (used so a later pick never re-selects an earlier column).
func columnIndexExcept(header []string, exclude []int, names ...string) int {
	for i, h := range header {
		if containsInt(exclude, i) {
			continue
		}
		hn := strings.ToLower(strings.Trim(h, "*_` "))
		for _, n := range names {
			if strings.Contains(hn, n) {
				return i
			}
		}
	}
	return -1
}

func containsInt(list []int, v int) bool {
	for _, x := range list {
		if x == v {
			return true
		}
	}
	return false
}

// ---------------------------------------------------------------------------
// Rendering
// ---------------------------------------------------------------------------

// RenderFlashcards renders a deck to a canonical flashcards.md document. It is
// the inverse of ParseFlashcards for the two data tables (the skill's prose
// sections are not reproduced — the writer's mutation PATH instead edits the
// existing document in place; see ApplyReview).
func RenderFlashcards(deck Deck) string {
	var b strings.Builder
	b.WriteString("# Flashcards\n\n")
	b.WriteString("## Cards\n\n")
	b.WriteString(joinCells([]string{"ID", "Front", "Back", "Anchor", "Tag", "Stage"}) + "\n")
	b.WriteString("|---|---|---|---|---|---|\n")
	for _, c := range deck.Cards {
		b.WriteString(joinCells([]string{c.ID, c.Front, c.Back, c.Anchor, c.Tag, string(c.Stage)}) + "\n")
	}
	b.WriteString("\n## Review log\n\n")
	b.WriteString(joinCells([]string{"ID", "Date", "Grade", "Next due"}) + "\n")
	b.WriteString("|---|---|---|---|\n")
	for _, r := range deck.Reviews {
		b.WriteString(joinCells([]string{r.ID, r.Date, string(r.Grade), r.NextDue}) + "\n")
	}
	return b.String()
}

// ---------------------------------------------------------------------------
// Scheduling (mirrors frontend lib/spacedRepetition.ts)
// ---------------------------------------------------------------------------

// IntervalScheduleDays is the skill's default interval ladder, in days.
var IntervalScheduleDays = []int{1, 3, 7, 16, 35}

// NewIntervalIndex is a never-reviewed card (not yet scheduled).
const NewIntervalIndex = -1

// ReviewState is a card's scheduling state: its progress rung plus the derived
// stage.
type ReviewState struct {
	IntervalIndex int            `json:"interval_index"`
	Stage         FlashcardStage `json:"stage"`
}

func clampIndex(index int) int {
	if index < NewIntervalIndex {
		return NewIntervalIndex
	}
	if index > len(IntervalScheduleDays)-1 {
		return len(IntervalScheduleDays) - 1
	}
	return index
}

// StageForIndex derives the stage implied by an interval index.
func StageForIndex(index int) FlashcardStage {
	i := clampIndex(index)
	switch {
	case i < 0:
		return StageNew
	case i == 0:
		return StageLearning
	default:
		return StageReview
	}
}

// StateForIndex builds a review state from an interval index.
func StateForIndex(index int) ReviewState {
	i := clampIndex(index)
	return ReviewState{IntervalIndex: i, Stage: StageForIndex(i)}
}

// IntervalDays returns the interval length (days) for an index, clamped
// (a never-reviewed card maps to the first rung).
func IntervalDays(index int) int {
	i := clampIndex(index)
	if i < 0 {
		i = 0
	}
	return IntervalScheduleDays[i]
}

// AdvanceIndex applies a grade to the current rung: `again` resets to the first
// rung, `hard` holds, `good`/`easy` advance one (capped). A never-reviewed card
// lands on the first rung for every grade.
func AdvanceIndex(currentIndex int, grade Grade) int {
	current := clampIndex(currentIndex)
	switch grade {
	case GradeAgain:
		return 0
	case GradeHard:
		if current < 0 {
			return 0
		}
		return current
	case GradeGood, GradeEasy:
		if current < 0 {
			return 0
		}
		if current+1 > len(IntervalScheduleDays)-1 {
			return len(IntervalScheduleDays) - 1
		}
		return current + 1
	default:
		return current
	}
}

// StateFromReviews replays a card's review log to recover its current state.
func StateFromReviews(entries []ReviewEntry) ReviewState {
	index := NewIntervalIndex
	for _, e := range entries {
		if e.Grade == "" {
			continue
		}
		index = AdvanceIndex(index, e.Grade)
	}
	return StateForIndex(index)
}

// ScheduleAfter grades a card at state on date, returning the resulting state
// and the computed next due date (YYYY-MM-DD).
func ScheduleAfter(state ReviewState, grade Grade, date string) (next ReviewState, nextDue string) {
	index := AdvanceIndex(state.IntervalIndex, grade)
	return StateForIndex(index), addDays(date, IntervalDays(index))
}

// addDays adds days to a YYYY-MM-DD date; an unparseable date is returned
// unchanged.
func addDays(date string, days int) string {
	t, err := time.Parse("2006-01-02", strings.TrimSpace(date))
	if err != nil {
		return date
	}
	return t.AddDate(0, 0, days).Format("2006-01-02")
}

// ---------------------------------------------------------------------------
// Surgical in-place update (pure)
// ---------------------------------------------------------------------------

// tableBlock is a contiguous run of Markdown table lines, half-open [start, end).
type tableBlock struct{ start, end int }

func tableBlocks(lines []string) []tableBlock {
	var blocks []tableBlock
	for i := 0; i < len(lines); {
		if !isTableRow(cleanLine(lines[i])) {
			i++
			continue
		}
		start := i
		for i < len(lines) && isTableRow(cleanLine(lines[i])) {
			i++
		}
		if _, ok := buildTable(lines[start:i]); ok {
			blocks = append(blocks, tableBlock{start: start, end: i})
		}
	}
	return blocks
}

// ApplyReview edits a flashcards.md document in place: it appends a review-log
// row for cardID graded on date and updates that card's Stage cell, leaving all
// other content (including the skill's prose) byte-for-byte intact. It fails
// closed when the deck has no card row for cardID or no review-log table.
func ApplyReview(content, cardID string, grade Grade, date string) (string, error) {
	if strings.TrimSpace(cardID) == "" {
		return "", errors.New("flashcard review rejected: empty card id")
	}
	if NormalizeGrade(string(grade)) == "" {
		return "", fmt.Errorf("flashcard review rejected: unknown grade %q", grade)
	}
	// Canonicalize the grade so the appended row is always well-formed.
	grade = NormalizeGrade(string(grade))

	lines := strings.Split(content, "\n")
	var cardsBlock, reviewBlock *tableBlock
	for _, b := range tableBlocks(lines) {
		header := splitCells(cleanLine(lines[b.start]))
		if cardsBlock == nil && isCardsHeader(header) {
			bb := b
			cardsBlock = &bb
			continue
		}
		if reviewBlock == nil && isReviewHeader(header) {
			bb := b
			reviewBlock = &bb
		}
	}
	if cardsBlock == nil {
		return "", errors.New("flashcard review rejected: no card table found")
	}
	if reviewBlock == nil {
		return "", errors.New("flashcard review rejected: no review-log table found")
	}

	// Locate the card row and its stage column.
	header := splitCells(cleanLine(lines[cardsBlock.start]))
	idIdx := colOr(header, 0, "id")
	stageIdx := columnIndex(header, "stage", "state", "status")
	cardRow := -1
	for i := cardsBlock.start + 2; i < cardsBlock.end; i++ {
		if cellAt(splitCells(cleanLine(lines[i])), idIdx) == cardID {
			cardRow = i
			break
		}
	}
	if cardRow < 0 {
		return "", fmt.Errorf("flashcard review rejected: card %q not found in the deck", cardID)
	}

	// Replay the card's log to schedule the new review.
	entries := reviewsForCardID(ParseFlashcards(content).Reviews, cardID)
	next, nextDue := ScheduleAfter(StateFromReviews(entries), grade, date)

	// Build the updated card row (only when the table carries a stage column).
	newCardLine := lines[cardRow]
	if stageIdx >= 0 {
		cells := splitCells(cleanLine(lines[cardRow]))
		for len(cells) <= stageIdx {
			cells = append(cells, "")
		}
		cells[stageIdx] = string(next.Stage)
		newCardLine = joinCells(cells)
	}

	// Build the review row against the log's own header width.
	rHeader := splitCells(cleanLine(lines[reviewBlock.start]))
	row := make([]string, len(rHeader))
	set := func(names []string, fallback int, value string) {
		idx := colOr(rHeader, fallback, names...)
		if idx >= 0 && idx < len(row) {
			row[idx] = value
		}
	}
	set([]string{"id"}, 0, cardID)
	set([]string{"date", "when"}, 1, date)
	set([]string{"grade", "rating", "result"}, 2, string(grade))
	set([]string{"next", "due"}, 3, nextDue)
	newReviewLine := joinCells(row)

	out := make([]string, 0, len(lines)+1)
	for i, line := range lines {
		if i == reviewBlock.end {
			out = append(out, newReviewLine)
		}
		if i == cardRow {
			out = append(out, newCardLine)
			continue
		}
		out = append(out, line)
	}
	if reviewBlock.end >= len(lines) {
		out = append(out, newReviewLine)
	}
	return strings.Join(out, "\n"), nil
}

func reviewsForCardID(all []ReviewEntry, cardID string) []ReviewEntry {
	var out []ReviewEntry
	for _, e := range all {
		if e.ID == cardID {
			out = append(out, e)
		}
	}
	return out
}

File: frontend/src/lib/flashcards.ts
// Flashcards parser for the study-paper skill's `flashcards.md` artifact.
//
// The skill writes a deck as a Markdown document (see
// core/papers/skills/study-paper/assets/flashcards.md) with two data tables:
//
//   ## 2. Cards        | ID | Front (question) | Back (answer) | Anchor | Tag | Stage |
//   ## 3. Review log   | ID | Date | Grade | Next due |
//
// The plain Markdown tables carry the review state (the append-only grade
// history, which the reader replays to derive each card's stage) that the
// backend's structured PaperRecord does NOT capture — PaperRecord models the
// card, not the deck — so the reader re-parses the raw artifact here to drive
// interactive review. The cards table's Stage column is write-only from this
// module's perspective: the stage is DERIVED from the review log (see
// lib/spacedRepetition.ts) and never read from the parsed card.
//
// Everything in this module is PURE over strings — no React, no DOM, no I/O —
// so the deck extraction is unit-testable on fixtures. Table parsing is shared
// with the critical-layer widgets (paperWidgets.parseMarkdownTables) so both
// mirror core/papers/parser.go identically.

import { parseMarkdownTables, type MarkdownTable } from './paperWidgets'

/** A card's lifecycle stage (mirrors the skill's `new` / `learning` / `review`). */
export type FlashcardStage = 'new' | 'learning' | 'review'

/** A self-graded review outcome. */
export type FlashcardGrade = 'again' | 'hard' | 'good' | 'easy'

/** One atomic question/answer card. */
export interface Flashcard {
  /** Stable deck-local label (e.g. `P1-01`). */
  id: string
  /** The question / prompt. */
  front: string
  /** The answer / fact. */
  back: string
  /** Where in the paper the fact comes from (`[Analyst]` for the reader's own synthesis). */
  anchor: string
  /** Topic / concept tag, for filtering. */
  tag: string
}

/** One row of the review log. */
export interface ReviewEntry {
  /** The card the grade belongs to. */
  id: string
  /** The review date (verbatim, typically `YYYY-MM-DD`). */
  date: string
  /** The self-grade; '' when the row carries an unrecognized/blank grade. */
  grade: FlashcardGrade | ''
  /** The next due date the grade produced (verbatim). */
  nextDue: string
}

/** A parsed deck: the ordered cards plus the append-only review log. */
export interface FlashcardDeck {
  cards: Flashcard[]
  reviews: ReviewEntry[]
}

// --- Column resolution (case-insensitive, markdown-emphasis tolerant) ---

function norm(s: string): string {
  return s
    .toLowerCase()
    .replace(/[*_`]/g, '')
    .trim()
}

/**
 * Normalize a cell VALUE (as opposed to a header): Markdown emphasis and
 * surrounding punctuation are stripped, then lower-cased and trimmed. Mirrors
 * the Go parser's shared value-normalizer, so a `**review**` or "`good`." cell
 * folds the same on both sides of the boundary.
 */
function normValue(s: string): string {
  return norm(s).replace(/^\.+|\.+$/g, '')
}

/** First header index matching any pattern, skipping already-used indices. */
function pickColumn(header: string[], patterns: RegExp, exclude: number[] = []): number {
  for (let i = 0; i < header.length; i++) {
    if (exclude.includes(i)) continue
    if (patterns.test(norm(header[i]!))) return i
  }
  return -1
}

function cell(row: string[], i: number): string {
  return i >= 0 && i < row.length ? (row[i] ?? '').trim() : ''
}

const ID_COL = /^id\b/
const FRONT_COL = /front|question|prompt/
const BACK_COL = /back|answer/
const ANCHOR_COL = /anchor|source|location|\bref\b|where|page/
const TAG_COL = /\btag\b|topic|concept|\blabel\b/
const DATE_COL = /date|when/
const GRADE_COL = /grade|rating|result/
const NEXT_DUE_COL = /next|\bdue\b/

/** Fold a raw stage token to its canonical value (unknown/blank → `new`).
 *  Emphasis/punctuation are stripped so a `**review**` cell folds like
 *  `review` — matching the Go parser's NormalizeStage. Kept as the documented
 *  cross-boundary mirror even though the deck's Stage column is derived from the
 *  review log and is no longer surfaced on the parsed card (see `Flashcard`). */
export function normalizeStage(raw: string): FlashcardStage {
  const s = normValue(raw)
  switch (s) {
    case 'learning':
    case 'learn':
      return 'learning'
    case 'review':
    case 'reviewing':
    case 'graduated':
      return 'review'
    case 'new':
      return 'new'
    default:
      return 'new'
  }
}

/** Fold a raw grade token to its canonical value (unknown/blank → '').
 *  Emphasis/punctuation are stripped so a "`good`" cell folds like `good` —
 *  matching the Go parser's NormalizeGrade. */
export function normalizeGrade(raw: string): FlashcardGrade | '' {
  const s = normValue(raw)
  switch (s) {
    case 'again':
    case 'hard':
    case 'good':
    case 'easy':
      return s
    default:
      return ''
  }
}

/** A table qualifies as the cards table when it carries a Front and a Back
 *  column (the deck's card table; its ID column is resolved separately so a
 *  header-only variant still yields cards). */
function isCardsTable(t: MarkdownTable): boolean {
  return pickColumn(t.header, FRONT_COL) >= 0 && pickColumn(t.header, BACK_COL) >= 0
}

/** A table qualifies as the review log when it carries a Grade column. */
function isReviewTable(t: MarkdownTable): boolean {
  return pickColumn(t.header, GRADE_COL) >= 0 || pickColumn(t.header, NEXT_DUE_COL) >= 0
}

function parseCardTable(t: MarkdownTable): Flashcard[] {
  const idIdx = pickColumn(t.header, ID_COL)
  const frontIdx = pickColumn(t.header, FRONT_COL)
  const backIdx = pickColumn(t.header, BACK_COL)
  const anchorIdx = pickColumn(t.header, ANCHOR_COL, [idIdx, frontIdx, backIdx])
  const tagIdx = pickColumn(t.header, TAG_COL, [idIdx, frontIdx, backIdx, anchorIdx])
  const cards: Flashcard[] = []
  for (const row of t.rows) {
    const id = cell(row, idIdx)
    // Card-row identity (shared with the Go parser's isCardRow): a data row is a
    // card unless the deck carries an id column and the row's id cell is empty.
    if (idIdx >= 0 && id === '') continue
    cards.push({
      id,
      front: cell(row, frontIdx),
      back: cell(row, backIdx),
      anchor: cell(row, anchorIdx),
      tag: cell(row, tagIdx),
    })
  }
  return cards
}

function parseReviewTable(t: MarkdownTable): ReviewEntry[] {
  const idIdx = pickColumn(t.header, ID_COL)
  const dateIdx = pickColumn(t.header, DATE_COL, [idIdx])
  const gradeIdx = pickColumn(t.header, GRADE_COL, [idIdx, dateIdx])
  const dueIdx = pickColumn(t.header, NEXT_DUE_COL, [idIdx, dateIdx, gradeIdx])
  const reviews: ReviewEntry[] = []
  for (const row of t.rows) {
    const id = cell(row, idIdx)
    const date = cell(row, dateIdx)
    const grade = cell(row, gradeIdx)
    const nextDue = cell(row, dueIdx)
    if (id === '' && date === '' && grade === '' && nextDue === '') continue
    reviews.push({ id, date, grade: normalizeGrade(grade), nextDue })
  }
  return reviews
}

/**
 * Parse a `flashcards.md` document into its deck. Best-effort and total: a
 * document with no recognizable tables yields an empty deck (the caller renders
 * the raw Markdown fallback), and rows that carry no card/grade are skipped.
 */
export function parseFlashcards(content: string): FlashcardDeck {
  const tables = parseMarkdownTables(content)
  const cardsTable = tables.find(isCardsTable)
  const reviewTable = tables.find((t) => t !== cardsTable && isReviewTable(t))
  return {
    cards: cardsTable ? parseCardTable(cardsTable) : [],
    reviews: reviewTable ? parseReviewTable(reviewTable) : [],
  }
}

/** The review entries belonging to one card, in log order. */
export function reviewsForCard(deck: FlashcardDeck, cardId: string): ReviewEntry[] {
  return deck.reviews.filter((entry) => entry.id === cardId)
}
